atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/watch/LockWatchTransactionManagerImpl.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.impl.watch;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Streams;
+import com.palantir.atlasdb.keyvalue.api.watch.ImmutableStartTransactionsResponse;
+import com.palantir.atlasdb.keyvalue.api.watch.LockWatchTransactionManager;
+import com.palantir.atlasdb.keyvalue.api.watch.StartTransactionRequest;
+import com.palantir.atlasdb.keyvalue.api.watch.StartTransactionsResponse;
+import com.palantir.atlasdb.transaction.api.AutoDelegate_TransactionManager;
+import com.palantir.atlasdb.transaction.api.OpenTransaction;
+import com.palantir.atlasdb.transaction.api.PreCommitCondition;
+import com.palantir.atlasdb.transaction.api.Transaction;
+import com.palantir.atlasdb.transaction.api.TransactionFailedRetriableException;
+import com.palantir.atlasdb.transaction.api.TransactionManager;
+import com.palantir.atlasdb.transaction.api.TransactionTask;
+import com.palantir.atlasdb.transaction.impl.ForwardingTransaction;
+import com.palantir.lock.watch.IdentifiedVersion;
+import com.palantir.lock.watch.LockWatchEventCache;
+import com.palantir.lock.watch.TransactionsLockWatchEvents;
+
+public class LockWatchTransactionManagerImpl implements AutoDelegate_TransactionManager, LockWatchTransactionManager {
+
+    private final TransactionManager transactionManager;
+    private final LockWatchEventCache lockWatchEventCache;
+
+    public LockWatchTransactionManagerImpl(
+            TransactionManager transactionManager,
+            LockWatchEventCache lockWatchEventCache) {
+        this.transactionManager = transactionManager;
+        this.lockWatchEventCache = lockWatchEventCache;
+    }
+
+    @Override
+    public TransactionManager delegate() {
+        return transactionManager;
+    }
+
+    @Override
+    public StartTransactionsResponse startTransactions(Optional<IdentifiedVersion> lastKnownVersion,
+            List<StartTransactionRequest> requests) {
+        List<CombinedCondition> amendedConditions = requests.stream().map(
+                request -> new CombinedCondition(request.preCommitCondition(),
+                        new WatchPreCommitCondition(lockWatchEventCache, request.watchCommitCondition())))
```

Comment:
this API is a bit goofy. The point of a PrecommitCondition is that you should just be able to pass it in, it should be entirely pluggable.

Here, the reason you need the extra method is because you need to get the start timestamp of the transactions that have been passed in.

This would be much easier if we just passed the precommit condition the start timestamp at some point.

Reply:
like, the rough minimal thing i'm advocating for here is that I should be able to write a

```java
interface CacheLocator extends PrecommitCondition {}

interface CacheFactory {
    Cache getCache(CacheLocator initializedLocator, Transaction transactionForFallbacks);
    CacheLocator newLocator();
}

and then use it like:

CacheLocator locator = cacheFactory.newLocator();
txnManager.runWithConditionWithRetry(locator, txn -> {
    Cache cache = cacheFactory.getCache(locator, txn);
});
```

Of course, the consumer doesn't use this API, but in Atlas it should read like this.